library/java/org/chromium/net/impl/CancelProofEnvoyStream.java

```diff
@@ -7,48 +7,58 @@
 import io.envoyproxy.envoymobile.engine.EnvoyHTTPStream;
 
 /**
- * Consistency layer above the {@link EnvoyHTTPStream} preventing unwarranted Stream operations
- * after a "cancel" operation. There are no "synchronized" - this is Compare And Swap based logic.
+ * CancelProofEnvoyStream is a consistency layer above the {@link EnvoyHTTPStream} preventing
+ * unwarranted Stream operations after a "cancel" operation. There are no "synchronized" - this is
+ * Compare And Swap based logic. This class is Thread Safe.
  *
  * <p>This contraption ensures that once a "cancel" operation is invoked, there will be no further
  * operations allowed with the EnvoyHTTPStream - subsequent operations will be ignored silently.
  * However, in the event that that one or more EnvoyHTTPStream operations are currently being
  * executed, the "cancel" operation gets postponed: the last concurrent operation will invoke
  * "cancel" at the end.
  *
- * <p>Instance of this class start with a state of "Busy Starting". This ensure that if a cancel
+ * <p>Instances of this class start with a state of "Busy Starting". This ensure that if a cancel
  * is invoked while the stream is being created, that cancel will be executed only once the stream
  * is completely initialized. Doing otherwise leads to unpredictable outcomes.
  */
 final class CancelProofEnvoyStream {
 
   private static final int CANCEL_BIT = 0x8000;
+  /**
+   * mState mainly maintains a counter of how many Stream operations are currently in-flight.
+   * However, when 15 is flipped (0x8000), it indicates that cancel operation hase been requested.
+   * If the counter is greater than zero, then that "cancel" operation is postponed until the last
+   * in-flight operation finishes, once the counter back to "zero". Then that last operation also
+   * invokes "cancel". On the other hand, if the counter is already "zero" when invoking "cancel",
+   * then it means that there are no in-flight operations: the "cancel" operation is immediately
+   * executed. Once the state is "canceled", any new stream operation is silently ignored.
+   */
   private final AtomicInteger mState = new AtomicInteger(1); // Busy starting.
   private volatile EnvoyHTTPStream mStream;                  // Cancel can come from any Thread.
 
   /** Sets the stream. Can only be invoked once. */
   void setStream(EnvoyHTTPStream stream) {
     assert mStream == null;
     mStream = stream;
-    if (cancelNeedsToBeInvoked()) {
+    if (decreaseConcurrentlyRunningStreamOperationsAndReturnTrueIfAwaitingCancel()) {
```

Comment:
It's not obvious to me why this decreases the running count. Else where in the class we do:
```
incrementCount();
doThing()
decrementCount();
```
But here we're just decrementing the count. Is this the 2nd half of some other operation? If so, can you add a comment which explains what the other half of this operation is?

Reply:
The other half is done by the Constructor: the counter starts with a value of "1' (not "0" as someone would expect.

Comment added, thanks.